collector/nfs_linux.go

```diff
@@ -14,164 +14,324 @@
-				value, "udp")
-
-			value, _ = strconv.ParseFloat(fields[2], 64)
-			ch <- prometheus.MustNewConstMetric(
-				nfsNetReadsDesc, prometheus.CounterValue,
-				value, "tcp")
-
-			value, _ = strconv.ParseFloat(fields[3], 64)
-			ch <- prometheus.MustNewConstMetric(
-				nfsNetConnectionsDesc, prometheus.CounterValue,
-				value, "tcp")
-		} else if fields := rpcLineRE.FindStringSubmatch(line); fields != nil {
-			value, _ := strconv.ParseFloat(fields[1], 64)
-			ch <- prometheus.MustNewConstMetric(
-				nfsRPCOperationsDesc,
-				prometheus.CounterValue, value)
-
-			value, _ = strconv.ParseFloat(fields[2], 64)
-			ch <- prometheus.MustNewConstMetric(
-				nfsRPCRetransmissionsDesc,
-				prometheus.CounterValue, value)
-
-			value, _ = strconv.ParseFloat(fields[3], 64)
-			ch <- prometheus.MustNewConstMetric(
-				nfsRPCAuthenticationRefreshesDesc,
-				prometheus.CounterValue, value)
-		} else if fields := procLineRE.FindStringSubmatch(line); fields != nil {
-			version := fields[1]
-			for procedure, count := range strings.Split(fields[2], " ") {
-				value, _ := strconv.ParseFloat(count, 64)
-				ch <- prometheus.MustNewConstMetric(
-					nfsProceduresDesc,
-					prometheus.CounterValue,
-					value,
-					version,
-					nfsProcedures[version][procedure])
-			}
-		} else if line != "" {
-			return errors.New("Failed to parse line: " + line)
-		}
-	}
+	c.updateNFSNetworkStats(ch, &stats.Network)
+	c.updateNFSClientRPCStats(ch, &stats.ClientRPC)
+	c.updateNFSRequestsv2Stats(ch, &stats.V2Stats)
+	c.updateNFSRequestsv3Stats(ch, &stats.V3Stats)
+	c.updateNFSRequestsv4Stats(ch, &stats.ClientV4Stats)
+
 	return nil
 }
+
+// updateNFSNetworkStats collects statistics for network packets/connections.
+func (c *nfsCollector) updateNFSNetworkStats(ch chan<- prometheus.Metric, s *nfs.Network) {
+	ch <- prometheus.MustNewConstMetric(c.nfsNetReadsDesc, prometheus.CounterValue,
+		float64(s.UDPCount), "udp")
+	ch <- prometheus.MustNewConstMetric(c.nfsNetReadsDesc, prometheus.CounterValue,
+		float64(s.TCPCount), "tcp")
+	ch <- prometheus.MustNewConstMetric(c.nfsNetConnectionsDesc, prometheus.CounterValue,
+		float64(s.TCPConnect), "tcp")
+}
+
+// updateNFSClientRPCStats collects statistics for kernel server RPCs.
+func (c *nfsCollector) updateNFSClientRPCStats(ch chan<- prometheus.Metric, s *nfs.ClientRPC) {
+	ch <- prometheus.MustNewConstMetric(c.nfsRPCOperationsDesc, prometheus.CounterValue,
+		float64(s.RPCCount))
+	ch <- prometheus.MustNewConstMetric(c.nfsRPCRetransmissionsDesc, prometheus.CounterValue,
+		float64(s.Retransmissions))
+	ch <- prometheus.MustNewConstMetric(c.nfsRPCAuthenticationRefreshesDesc, prometheus.CounterValue,
+		float64(s.AuthRefreshes))
+}
+
+// updateNFSRequestsv2Stats collects statistics for NFSv2 requests.
+func (c *nfsCollector) updateNFSRequestsv2Stats(ch chan<- prometheus.Metric, s *nfs.V2Stats) {
+	const proto = "2"
+	ch <- prometheus.MustNewConstMetric(c.nfsProceduresDesc, prometheus.CounterValue,
```

Comment:
Looks like these all use the same procedures descriptor.

Since most of these fields are identical, you may just want to go with a slice of anonymous structures for the descriptor, value, and label, and then iterate over it to perform the `MustNewConstMetric` call.

Reply:
Yes, that seems reasonable, but for now there's a bunch of exceptions to the label values.  How about we use the cleanup PR to do this?

Reply:
SGTM

Reply:
Hrmm, what stops you exactly from doing this right away? Doesn't look like there is anything dynamic that would prevent this, no?

Reply:
There are several not 1:1 mappings.  `FsStat` vs `statfs` for example.  Some splitting of `CamelCase` to `under_scores`.

Reply:
Yeah no need for magic, but can you do the onces that are 1:1 mappings in a loop?